MRTK SDK UX - ObjectCollection

radicalad opened the pull request:
Overview:
Porting ObjectCollection to the vNext branch. Allow users to easily arrange items in a grid in various volumes (planes, spheres, cylinders).


Changes:
Abstracted original ObjectCollection to a BaseCollection class. This makes for a good pattern to extend from in the future. Additionally, the custom editor script is now inherited by all children.


To use BaseCollection:
Inherit BaseCollection and override the abstract method LayoutChildren(). Easy as that.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <summary>
+        /// Action called when collection is updated
+        /// </summary>
+        public Action<BaseCollection> OnCollectionUpdated;
```
Should we update this to be an event routed through the MRTK?

keveleigh replied:
I'm not what we'd gain that'd be worth the perf impact of routing local events like this through the event system. I'm sure @radicalad knows more about the use of this, it seems to me like mostly the objects in the collection or related objects would care about this event, instead of routing it to every global/modal/etc listener as well.

StephenHodgson replied:
I was mostly just curious. I don't mind keeping the action, but also adding an event.

radicalad replied:
I'm kinda into it purely for an excuse to get to know the new system better. As far as use case, I get why its there in theory, but I've never used it myself, I almost scrapped it when I revisited it for this PR.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <summary>
+        /// Whether to treat inactive transforms as 'invisible'
+        /// </summary>
+        public bool IgnoreInactiveTransforms = true;
```
fields should not be public. Please update them to be private and serialized. If a field needs to be accessed outside of the scope of the class, please add a property accessor.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/LayoutTypeEnum.cs:
```diff
@@ -0,0 +1,16 @@
+    /// </summary>
+    public enum LayoutTypeEnum
+    {
+        ColumnThenRow,          // Sort by column, then by row
```
These comments could be summary tags instead.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/SurfaceTypeEnum.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    /// The type of surface to map the collect to.
+    /// </summary>
+    public enum SurfaceTypeEnum
```
Some of these enums should probably be defined in the mrtk core

radicalad replied:
Any guidance on organizational structure? _Core.Utilities.Enums? SortType and OrientType seem to be the most useful.

StephenHodgson replied:
`_Core.Definitions.Utilities`

StephenHodgson replied:
That's where most of our enums are defined.

radicalad replied:
Ah okay, the naming convention changed. Got it.

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+
+        public virtual void UpdateCollection()
+        {
+
```
nit: extra line

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionNode.cs:
```diff
@@ -0,0 +1,21 @@
+    /// Collection node is a data storage class for individual data about an object in the collection.
+    /// </summary>
+    [Serializable]
+    public class CollectionNode
```
Should this be a struct?

StephenHodgson replied:
If we end up changing this to a struct we should also move it to the core/definitions/utilities folder

radicalad replied:
I realized I have null checks for the CollectionNode list. I switched it back to a class.

StephenHodgson replied:
Why would a node be null?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+
+            for (int i = 0; i < NodeList.Count; i++)
+            {
+                if (NodeList[i].transform == null || (IgnoreInactiveTransforms && !NodeList[i].transform.gameObject.activeSelf) || NodeList[i].transform.parent == null || !(NodeList[i].transform.parent.gameObject == this.gameObject))
```
code style: I don't believe we have been using "this." in the toolkit.

radicalad replied:
Old habits die hard. 😄

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/Editor/BaseCollectionEditor.cs:
```diff
@@ -0,0 +1,26 @@
+using UnityEditor;
+
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Collections
```
`namespace Microsoft.MixedReality.Toolkit.SDK.UX.Inspectors.Collections`

Also, please update the editor folder name to be `Inspectors` instead.

StephenHodgson replied:
Folder location should also reflect the namespace I provided above.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/Editor/BaseCollectionEditor.cs:
```diff
@@ -0,0 +1,26 @@
+        public override void OnInspectorGUI()
+        {
+            // Draw the default
+            base.OnInspectorGUI();
```
Let's go ahead and grab all the properties off the object so we can have a clean inspector.

radicalad replied:
Not sure what you mean here. I'm just adding a custom button.

StephenHodgson replied:
Just grab all the serialized properties, and implement them in the gui update. 

If you'd like I can open a PR that does this, if it's still unclear.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridCollection.cs:
```diff
@@ -0,0 +1,451 @@
+        /// Type of surface to map the collection to.
+        /// </summary>
+        [Tooltip("Type of surface to map the collection to")]
+        public SurfaceTypeEnum SurfaceType = SurfaceTypeEnum.Plane;
```
I'm pretty sure some of these enums are already defined in core.

radicalad replied:
I didn't see any existing enums when I did a search -> related to the previous comment. Which should we pull in?

StephenHodgson replied:
No worries, I just wanted to double check. I couldn't find any either. The only one that looks similar is PivotAxis, but I think these area all good here.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Collections
+{
+    public abstract class BaseCollection : MonoBehaviour
```
Should we have a collection interface?

radicalad replied:
What are you thinking would be included in it vs the base class? LayoutChildren()?

StephenHodgson replied:
Anything that defines a collection.

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        /// <summary>
+        /// Internal function to check if a node exists in the NodeList.
```
Don't need to mention the function is internal. It is marked private

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseCollection.cs:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        /// <summary>
+        /// Protected Internal function for laying out all children when UpdateCollection is called.
```
Similar to the earlier comment. No need to specify the visibility in the comments

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/LayoutTypeEnum.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    /// Collection layout type enum
+    /// </summary>
+    public enum LayoutTypeEnum
```
Rename to ObjectCollectionLayoutType (or similar)?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/OrientTypeEnum.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// Orientation type enum for collections
+    /// </summary>
+    public enum OrientTypeEnum
```
ObjectCollectionOrientationType? Or, if more generic than that, OrientationType and move to a common location

StephenHodgson replied:
Sounds pretty generic.

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/SortTypeEnum.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// Sorting type for collections
+    /// </summary>
+    public enum SortTypeEnum
```
Rename to ObjectCollectionSortType?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/SurfaceTypeEnum.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    /// The type of surface to map the collect to.
+    /// </summary>
+    public enum SurfaceTypeEnum
```
ObjectOrientationSurfaceType?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionNode.cs:
```diff
@@ -0,0 +1,21 @@
+    /// Collection node is a data storage class for individual data about an object in the collection.
+    /// </summary>
+    [Serializable]
+    public class CollectionNode
```
ObjectCollectionNode?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/Editor/BaseCollectionEditor.cs:
```diff
@@ -0,0 +1,26 @@
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Collections
+{
+    [CustomEditor( typeof(BaseCollection), true )]
+    public class CollectionEditor : Editor
```
BaseObjectCollectionInspector?

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/Editor/BaseCollectionEditor.cs:
```diff
@@ -0,0 +1,26 @@
+using UnityEngine;
+using UnityEditor;
+
+
```
nit: extra line

davidkline-ms commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridCollection.cs:
```diff
@@ -0,0 +1,451 @@
+        /// Should the objects in the collection face the origin of the collection
+        /// </summary>
+        [Tooltip("Should the objects in the collection be rotated / how should they be rotated")]
+        public OrientTypeEnum OrientType = OrientTypeEnum.FaceOrigin;
```
code style: public property with private backing field

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -103,21 +117,19 @@ public virtual void UpdateCollection()
                     {
                         return true;
                     }
+
```
extra lines?

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <summary>
+        /// Protected Internal function for laying out all children when UpdateCollection is called.
+        /// </summary>
+        protected virtual void LayoutChildren() { }
```
abstract ?

radicalad replied:
good catch.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+        }
+
+        // Internal function for getting the relative mapping based on a source Vec3 and a radius for cylinder mapping.
+        // source to be mapped to cylinder, Radius for the radius of the cylinder
```
Should these be xml format?

```
/// <summary>
/// Comments
/// </summary>
```

radicalad replied:
@davidkline-ms marked a change to drop them since they're private/internal. I figured I'd leave the explanation - just removed the XML format.

davidkline-ms replied:
I was just looking for "Internal function for getting the ..." to be changed to "Get the ..."

There is no real need to specify the visibility in the comment (it's easy to see in the method signature :)

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+
+        // Internal function for getting the relative mapping based on a source Vec3 and a radius for spherical mapping.
+        // The source to be mapped to sphere, radius the radius of the sphere</param>
+        private Vector3 SphericalMapping(Vector3 source, float radius)
```
Seems like a nice vector 3 math extension...

radicalad commented:
Updates made. Left CollectionNode is a class -> had some value assignment issues with structs being immutable. Left the public Action alone as well as the protected virtual method LayoutChildren. The base class approach will give more for free (not needing to implement UpdateCollection() etc) vs an interface.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+        {
+            get { return height; }
+        }
+        #endregion
```
we've been adding the name of the region at the beginning and end to help identify them better.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+                    {
+                        newPos = nodeGrid[i];
+                        NodeList[i].transform.localPosition = newPos;
+                        switch (OrientType)
```
With all these nested switch statements, let's call a method instead and pass any values it needs to it.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+        }
+
+        // Internal function for randomized mapping based on a source Vec3 and a radius for randomization distance.
+        private Vector3 ScatterMapping(Vector3 source, float radius)
```
This could also be a handy vector 3 extension.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+        {
+            // Sort by closest to center (don't worry about z axis)
+            // Use the position of the collection as the packing center
+            nodes.Sort(delegate (ObjectCollectionNode circle1, ObjectCollectionNode circle2) {
```
let's drop the bracket onto the next line.

StephenHodgson replied:
Could we make this a method instead of a delegate or lamda?

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+                        {
+                            if (cellCounter < NodeList.Count)
+                            {
+                                nodeGrid[cellCounter] = new Vector3((c * CellWidth) - startOffsetX + halfcell.x, -(r * CellHeight) + startOffsetY - halfcell.y, 0f) + (Vector3)((NodeList[cellCounter])).Offset;
```
We probably don't need to assign a new vector here, seeing as we create a new grid array anyway.

StephenHodgson replied:
Just assign the x,y,z values

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,482 @@
+                        {
+                            if (cellCounter < NodeList.Count)
+                            {
+                                nodeGrid[cellCounter] = new Vector3((c * CellWidth) - startOffsetX + halfcell.x, -(r * CellHeight) + startOffsetY - halfcell.y, 0f) + (Vector3)((NodeList[cellCounter])).Offset;
```
We probably don't need to assign a new vector here, seeing as we create a new grid array anyway.
We should be able to just set the x,y,z values.

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/OrientationType.cs:
```diff
@@ -0,0 +1,32 @@
+        /// Zero rotation + 180 degrees
+        /// </summary>
+        FaceForwardReversed,
+    }
```
Erm, it looks like this is not being ported from the latest version of the object collection. Here are very significant changes I made for the June/July release:
https://github.com/Ecnassianer/MixedRealityToolkit-Unity/commit/945401107c48493cd1e649c5a92090cf4710a7b8#diff-67828c8d64af75261de12e4344488104

Ecnassianer commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/CollectionEnums/ObjectOrientationSurfaceType.cs:
```diff
@@ -11,6 +11,7 @@ public enum ObjectOrientationSurfaceType
         Sphere,
-        Scatter,
+        Radial,
+        Scatter
```
Was this reordering intentional? I generally try to avoid renumbering enums when I add a new one, in part because it breaks existing serialized data. But more generally a philosophy of "if it's not broken, don't touch it". :)

radicalad replied:
Works for me,  I put it at the bottom of the list for preference.

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/OrientationType.cs:
```diff
@@ -28,5 +28,22 @@ public enum OrientationType
+        /// <summary>
+        /// Parent Relative Up
+        /// </summary>
+        FaceParentUp,
```
We should take the htk->vNext port as an opportunity to rename FaceForward and FaceForwardReversed to FaceParentForward and FaceParentBack.

Ecnassianer replied:
bump

radicalad replied:
Sure. Done.

radicalad commented on Assets/MixedRealityToolkit-Examples/Demos/Boundary/Scripts/BoundaryVisualizationDemo.cs:
```diff
@@ -5,6 +5,7 @@
 using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Boundary;
 using Microsoft.MixedReality.Toolkit.Internal.Interfaces.BoundarySystem;
 using Microsoft.MixedReality.Toolkit.Internal.Managers;
+using Microsoft.MixedReality.Toolkit.Internal.Utilities;
```
They showed up when I did a merge with the latest mrtk_development branch.

Ecnassianer replied:
I think the best way to not have this happen is this:
1) Always keep your fork's mrtk_development free of any changes.
2) When you want to update, merge Microsoft's mrtk_development into your fork's, and it will automatically fast forward with no merge to deal with. Beautiful clean.
3) Create your changes in a branch based on your fork's mrtk_development, don't ever* merge into it.
4) Whenever you get an update from Microsoft's fork, *rebase* your branch to the new tip of your fork's updated mrtk_development branch.

radicalad replied:
Yep, that's what I followed, not sure why all of this stuff showed up. I'll try to revert and see what I can do.

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/EventDatum/Boundary/BoundaryEventData.cs:
```diff
@@ -11,15 +11,6 @@ namespace Microsoft.MixedReality.Toolkit.Internal.EventDatum.Boundary
-        ///// </remarks>
-        // todo: coming in Beta
-        // public bool IsCeilingVisualized { get; private set; }
-
```
I think perhaps this should not appear in this diff?

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Utilities/MathUtilities.cs:
```diff
@@ -439,5 +439,19 @@ public static float RadiansToDegrees(float radians)
+        /// <returns>
+        /// The angle between the two points.
+        /// </returns>
+        public static float GetAngleBetween(Vector2 pointA, Vector2 pointB)
```
Is this different from Vector2.Angle? https://docs.unity3d.com/ScriptReference/Vector2.Angle.html If so, include the difference in the summary text, since it's not immediately apparent to me what the difference is.

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/SortType.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <summary>
+    /// Sorting type for collections
+    /// </summary>
+    public enum SortType
```
This is too generic of a name. "SortType" seems like it should have entries for QuickSort, BubbleSort, MergeSort etc. This should be something like:
3DSortType
SpatialSortType
VolumetricSortType
PositionalSortingType
GameObjectSortType

radicalad replied:
How About CollationOrderType. Technically this isn't for sorting - you need an order in which to sort.

Ecnassianer replied:
That feels like a synonym for "sort" to me? And it doesn't actually clarify what these are used for. =/

Maybe we're barking at the wrong tree by calling it a sort. Should it be something like SpatialLayoutType instead? SpatialPatternType? 3DLayoutType?

radicalad replied:
This actually has nothing to do with spatial layout. I'd suggest sticking with Collation here as it has to do with naming of objects vs when they appear in the hierarchy, conceptually it makes sense.

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/SortType.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <summary>
+    /// Sorting type for collections
+    /// </summary>
+    public enum SortType
```
Does vNext end enum names with "Enum"? i.e, SortTypeEnum. I've seen both in htk (SortType and SortTypeEnum), but I'm not sure what the guideline is for vNext.

StephenHodgson replied:
Type is the suffix we've been going with. But I'm fine with either as long as we make a decision.

radicalad replied:
Thats what I thought too, but going through everything, I didn't find any enum suffixes so i've since renamed everything. Personally, I feel having the suffix helps with self-documentation.

Ecnassianer replied:
I think this now has both type suffixes XD

radicalad commented:
Okay, sorry about the crazy merge history. We should be good now. I might have missed something as I rolled WAY back to get to a clean spot. Notable change: added the ...Enum suffix to any enum I'm modifying or adding.

Ecnassianer commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+            switch (SurfaceType)
+            {
+                case ObjectOrientationSurfaceTypeEnum.Plane:
+                    for (int i = 0; i < NodeList.Count; i++)
```
Would this inheritance be more useful if each SurfaceType called an overridable function? That way you would just override the behaviors you wanted differently, and could leave the rest of the LayoutChildren() function alone?

Ecnassianer replied:
Just seems like this is a huge block of functionality to override. :) (I am lazy)

radicalad replied:
That assumes SurfaceType needs to be implemented - which it doesn't. This Collection just happens to implement LayoutChildren with Surface types. The BaseObjectCollection contains the abstract LayoutChildren method - so more likely you would be inheriting the base collection and writing from scratch - not the grid collection. This is actually the main update I wanted to get into MRTK, as it is, the ObjectCollection is a great control, but since all of its specific functionality is embedded in the single class, its not very extensible.

Ecnassianer replied:
Aha! This all click into sense now that we've talked!

Ecnassianer commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+                    }
+
+                    // Iterate [x] times
+                    // TODO move center, iterations and padding into a public field
```
Is this TODO still relevant? I'm not sure what it's referring to

Ecnassianer commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/CollationOrderTypeEnum.cs:
```diff
@@ -0,0 +1,32 @@
+        /// </summary>
+        None,
+        /// <summary>
+        /// Sort by transform order
```
I haven't used this, so I'm not sure what "transform order" is, but this documentation doesn't help me understand. If you know what this order does, can you improve the summary?

radicalad replied:
I'm not sure, I've never used it before - this is leftover from the original enum.

radicalad replied:
It compares sibling index, I'll rename to better describe the behavior.

radicalad commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+
+        #region private fields
+        /// <summary>
+        /// Whether to treat inactive transforms as 'invisible'
```
Agreed.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <summary>
+        /// Action called when collection is updated
+        /// </summary>
+        public Action<BaseObjectCollection> OnCollectionUpdated;
```
Needs to be a property if public

StephenHodgson replied:
Where are we even creating and assigning this action?

radicalad replied:
in the UpdateCollection() method. This action gets called every time a user presses the custom inspector button. Like I said earlier, I haven't actually ever subscribed to this action in practice.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        /// List of objects with generated data on the object.
+        /// </summary>
+        [HideInInspector]
+        public List<ObjectCollectionNode> NodeList = new List<ObjectCollectionNode>();
```
Also needs to be a property.

radicalad replied:
Does List<> behave as expected when wrapped in a property? I have the change in a commit -> just not sure about the setter working as one would expect.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        #endregion
+
+        #region public accessors
+        public bool IgnoreInactiveTransforms
```
missing summary tags

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        #endregion
+
+
+        public virtual void UpdateCollection()
```
missing summary tags

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+        {
+            get { return surfaceType; }
+            set { surfaceType = value; }
+
```
extra space

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+    {
+        #region private fields
+        [Tooltip("Type of surface to map the collection to")]
+        private ObjectOrientationSurfaceTypeEnum surfaceType = ObjectOrientationSurfaceTypeEnum.Plane;
```
properties and their corresponding fields should be co-located.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+                    if (cellCounter < NodeList.Count)
+                    {
+                        grid[cellCounter].Set(
+                            ((i * CellWidth) - offsetX + halfCell.x) + NodeList[cellCounter].Offset.x, (-(j * CellHeight) + offsetY - halfCell.y) + NodeList[cellCounter].Offset.y, 0.0f);
```
Let's put each x,y,z value assignments on new lines here. Might make it a bit easier to read.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+            switch (OrientType)
+            {
+                case OrientationTypeEnum.FaceOrigin:
+                    node.transform.rotation = Quaternion.LookRotation(node.transform.position - this.transform.position, this.transform.up);
```
We can probably remove all the redundant `this`'s as well.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,446 @@
+        /// <param name="node"></param>
+        /// <param name="orientType"></param>
+        /// <param name="newPos"></param>
+        private void UpdateNodeFacing(ObjectCollectionNode node, OrientationTypeEnum orientType, Vector3 newPos = default(Vector3))
```
why not just vector3.zero?

Ecnassianer replied:
Dejavu :D 
https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2229#discussion_r192543633

StephenHodgson replied:
Seems as though orientType and newPos are unused in the method. We can just remove them altogether.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/ObjectCollectionNode.cs:
```diff
@@ -0,0 +1,21 @@
+    /// Collection node is a data storage class for individual data about an object in a collection.
+    /// </summary>
+    [Serializable]
+    public class ObjectCollectionNode
```
pretty sure we can probably make this a struct. Where are we checking if it's null?

radicalad replied:
Its not the null check (it was unnecessary so I removed it), ObjectCollectionNode needs to be mutable. Switching to a struct broke basically every line where properties were being modified.

StephenHodgson replied:
All you'd need to do is get a local copy, assign the values, then assign your modified copy back to the array. Like so:

```
case ObjectOrientationSurfaceTypeEnum.Scatter:
// Get randomized planar mapping
// Calculate radius of each node while we're here
// Then use the packer function to shift them into place
for (int i = 0; i < NodeList.Count; i++)
{
    var node = NodeList[i];
    newPos = VectorExtensions.ScatterMapping(nodeGrid[i], Radius);
    var nodeCollider = node.transform.GetComponentInChildren<Collider>();

    if (nodeCollider != null)
    {
        // Make the radius the largest of the object's dimensions to avoid overlap
        Bounds bounds = nodeCollider.bounds;
        node.Radius = Mathf.Max(Mathf.Max(bounds.size.x, bounds.size.y), bounds.size.z) * 0.5f;
    }
    else
    {
        // Make the radius a default value
        // TODO move this into a public field ?
        node.Radius = 1f;
    }

    node.transform.localPosition = newPos;
    UpdateNodeFacing(node);
    NodeList[i] = node;
}
```

radicalad replied:
Ah, thanks @StephenHodgson. 😄 Just curious, is there a perf reason for switching to a struct?

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Inspectors/Collections/BaseCollectionEditor.cs:
```diff
@@ -0,0 +1,26 @@
+        public override void OnInspectorGUI()
+        {
+            // Draw the default
+            base.OnInspectorGUI();
```
Let's go ahead and grab all the properties and show them here so we can a nice clean inspector.

radicalad replied:
I'm not sure what you mean by this? I don't typically do custom inspectors.

StephenHodgson replied:
If you'd like I can open a PR with the suggested changes. Shouldn't take but a few minutes.

Ecnassianer replied:
I think it is much better to rely on the base's default behavior. By overriding that behavior and rebuilding it here, we just make it harder to add properties to the class (e.g, they now need to be added in two places to show up in the inspector). From past discussions, I think @keveleigh is of the same opinion.

StephenHodgson replied:
Let's not be lazy and do it right.

radicalad replied:
Sure. whatever is easiest, is there an existing example somewhere in MRTK?

Ecnassianer replied:
What I'm asserting is that the base.OnInspectorGUI() or DrawDefaultInspector() methods is "doing it right". It removes code duplication and creates a clearer and easier to maintain inspector for future developers.

Think of it in a more generalized way: say we had a class that provided a virtual DrawMe() function, and then we got a PR to bring in a child class that overrode DrawMe() by copy pasting the body of the parent class's DrawMe() and then added a text label at the bottom. We'd ask for it to be refactored to a call to the parent function followed by adding a the text label.

This is the same, we're inheriting behavior from the default inspector. It's bad practice to duplicate that behavior.

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Extensions/VectorExtensions.cs:
```diff
@@ -161,5 +161,86 @@ public static bool IsValidPosition(this Vector3 position)
+        {
+            float circ = 2f * Mathf.PI * radius;
+
+            Vector3 newPos = new Vector3(0f, 0f, radius);
```
why make a new vector here? couldn't we just assign it back to the source?

Ecnassianer replied:
It looks like some of Kurtis' changes got lost in the port, you can see this was fixed in the RadialMapping() function in ObjectCollection.cs here:
https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2246/files/64f46d85b93b59faf2fd93f69a18deb9f706cdb0

StephenHodgson commented on Assets/MixedRealityToolkit/_Core/Extensions/VectorExtensions.cs:
```diff
@@ -161,5 +161,86 @@ public static bool IsValidPosition(this Vector3 position)
+        public static Vector3 CylindricalMapping(Vector3 source, float radius)
+        {
+            float circ  = 2f * Mathf.PI * radius;
+            Vector3 newPos = new Vector3(0f, source.y, radius);
```
same thing here. We probably don't need a new vector and can just update the source, then return it.

Ecnassianer replied:
I didn't bother to find this line in the previous PR, but this was probably also fixed there. (See my comment above)

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+        /// <summary>
+        /// Protected Internal function for laying out all children when UpdateCollection is called.
+        /// </summary>
+        protected abstract void LayoutChildren()
```
missing a `;`

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+                    break;
+
+                case CollationOrderTypeEnum.Transform:
+                    NodeList.Sort(delegate (ObjectCollectionNode c1, ObjectCollectionNode c2) { return c1.transform.GetSiblingIndex().CompareTo(c2.transform.GetSiblingIndex()); });
```
These could be lambda expressions. might shorten it down a bit and make it a tad easier to read.

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,143 @@
+                    break;
+
+                case CollationOrderTypeEnum.Alphabetical:
+                    NodeList.Sort(delegate (ObjectCollectionNode c1, ObjectCollectionNode c2) { return c1.Name.CompareTo(c2.Name); });
```
let's be sure to use ordinal comparison here. and on the apha reversed as well.

StephenHodgson commented:
Let's be sure to make all the icons the same for the collection components.
![image](https://user-images.githubusercontent.com/13334553/44371332-39b34d80-a493-11e8-8aab-26427c7aa301.png)

StephenHodgson commented:
Also please be sure to assign the editor inspector icon and rename this class to `BaseCollectionInspector`
![image](https://user-images.githubusercontent.com/13334553/44371355-5485c200-a493-11e8-8524-82fcdc27fda1.png)

radicalad commented:
Okay, I think I addressed everything. @StephenHodgson as far as the icon inconsistency, it makes sense to me for the actual UX component script to have the UX icon. Everything else is a back end script (and has the script icon). but the GridObjectCollection script is what you actually apply in a unity scene. I would expect the icon to reflect that.

SimonDarksideJ commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,419 @@
+    {
+        #region public accessors
+        [Tooltip("Type of surface to map the collection to")]
+        private ObjectOrientationSurfaceTypeEnum surfaceType = ObjectOrientationSurfaceTypeEnum.Plane;
```
Please ensure you use the correct spacings in the script. 
Properties both public and private should be spaced between them.

SimonDarksideJ commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,419 @@
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Don't use Throw, as it can case Unity to crash which makes it really hard to debug.

Use 
```Debug.LogWarning`` or ```Debug.LogError`` instead (depending on the seriousness of the error.
Include help text in the error where possible to ease diagnosis

SimonDarksideJ commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/CollationOrderTypeEnum.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// Don't sort, just display in order received
+        /// </summary>
+        None,
```
Initialise with
```None = 0,```
As per the enum standards.

SimonDarksideJ commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/OrientationTypeEnum.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <summary>
+        /// Don't rotate at all
+        /// </summary>
+        None,
```
Initialise with
```None = 0,```
As per the enum standards.

davidkline-ms commented on Assets/MixedRealityToolkit/_Core/Definitions/Utilities/OrientationTypeEnum.cs:
```diff
@@ -0,0 +1,50 @@
+        /// Lay flat on the surface, facing out
+        /// </summary>
+        FaceCenterAxisReversed,
+
```
nit: extra spaces

radicalad commented:
@SimonDarksideJ added a demo scene! Hopefully we're good to go. Your columns warning was actually a symptom of a larger bug - at some point I deleted the line that sets columns, kinda problematic considering its required for layout... 😄

SimonDarksideJ commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -149,8 +175,8 @@ public Mesh CylinderMesh
-
-        private float circumference;
+        
+        //private float circumference;
```
If it's not needed, then simply remove it (if you need it later, GIT will remember :D)

radicalad commented:
@Ecnassianer scatter is arguably still a "grid" layout, but its always felt like the odd man out. I've gone ahead and moved it to its own script, to minimize the amount of refactoring and duplicate code, I have it extending the *GridObjectCollection* and changed a few of the access modifiers in *GridObjectCollection* to make that work. Other than that, I'm with you on the suffixes, I prefer *enum* over *type*.

radicalad commented:
@Ecnassianer I just reread your comment - you're suggesting *type* over *enum*! @davidkline-ms @StephenHodgson @SimonDarksideJ what is the call here? I'm happy to go with either.

davidkline-ms commented:
> How come you approved this @davidkline-ms with a reported warning and your "nit" comment?
Especially can't have PR's generating warnings in the project at this stage

because an extra blank line or two is a nit-pick and not, in my book, a merge blocker.  I didn't notice the missing initialization, besides that is far less of a concern than breaking changes at this point in the project.

keveleigh commented:
@radicalad We've typically been using `Type` as the suffix for enums in vNext. Not sure where/when/if an official decision was made (@davidkline-ms?), but I believe that's been case.

radicalad commented:
'Type' it is then. Just want to be consistent.

davidkline-ms commented:
> Not sure where/when/if an official decision was made (@davidkline-ms?), but I believe that's been case.

I honestly do not recall. I believe we are not 100% consistent (not all enums have Type or any other suffix)

Ecnassianer commented:
Seems like this thread has critical mass to make an official decision?
Enum Suffix: 0
Type Suffix: 5 (David, Kurtis, Eric, Stephen Addison)

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/ScatterObjectCollection.cs:
```diff
@@ -0,0 +1,116 @@
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Collections
+{
+
+
```
missing summary

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/ScatterObjectCollection.cs:
```diff
@@ -0,0 +1,116 @@
+
+            // First start with a grid then project onto surface
+            ResolveGridLayout( nodeGrid, startOffsetX, startOffsetY, Layout );
+
```
extra line

StephenHodgson commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/BaseObjectCollection.cs:
```diff
@@ -0,0 +1,157 @@
+        }
+        #endregion public members
+
+
```
extra line

StephenHodgson commented:
Just some nitpicky stuff but looks good.

radicalad commented on Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs:
```diff
@@ -0,0 +1,377 @@
+                    if (cellCounter < NodeList.Count)
+                    {
+                        grid[cellCounter].Set
+                            (
```
Is there a convention I should be following?

StephenHodgson replied:
Yeah generally `(`'s are on the same line, but parameters can break.
